packages/components/src/ui/heading/test/index.js

```diff
@@ -0,0 +1,47 @@
+/**
+ * External dependencies
+ */
+import { render } from '@testing-library/react';
+
+/**
+ * Internal dependencies
+ */
+import { Heading } from '../index';
+
+describe( 'props', () => {
+	let base;
+	beforeEach( () => {
+		base = render( <Heading>Code is Poetry</Heading> );
+	} );
+
+	test( 'should render correctly', () => {
+		expect( base.container.firstChild ).toMatchSnapshot();
+	} );
+
+	test( 'should render level as a number', () => {
+		const { container } = render(
+			<Heading level={ 4 }>Code is Poetry</Heading>
+		);
+		expect( container.firstChild ).toMatchStyleDiffSnapshot(
+			base.container.firstChild
+		);
+	} );
+
+	test( 'should render level as a string', () => {
+		const { container } = render(
+			<Heading level="4">Code is Poetry</Heading>
+		);
+		expect( container.firstChild ).toMatchStyleDiffSnapshot(
+			base.container.firstChild
+		);
+	} );
+
+	test( 'should allow as prop', () => {
+		const { container } = render(
+			<Heading level="1" as="span">
+				Code is Poetry
+			</Heading>
+		);
+		expect( container.firstChild.tagName ).toBe( 'SPAN' );
+	} );
```

Comment:
When do you think rendering this component as another element would be the case? This looks like an a11y problem as the element would be styled as a heading but would lose its semantics.

If there are valid cases to accept elements other than the `h*` ones, we should add these attributes to maintain semantics:
```html
<span role="heading" aria-level="1">
```

Reply:
Okay! I can add that to the hook 🙂

Reply:
So I've gone ahead and added support for the properties you mentioned. That being said, I think the main reason to support elements other than the semantic ones here is just to maintain uniformity throughout the library of polymorphism. So far every component except `Portal` includes the `as` prop and they all behave the same way. I think we could probably exclude it from certain components but it would take some fiddling with the type system that is difficult. Maybe if we decide to exclude the `as` prop from this component we can explore that in the future?

Reply:
> Maybe if we decide to exclude the as prop from this component we can explore that in the future?

From my experience re: heading styles vs. tagNames.

This conflict usually exists when you need the semantics of an `h*` (e.g. `h2`), but the aesthetic sizing of another (e.g. `h4`).  For better or for worse, this workflow can be seen in many [Bootstrap based](https://getbootstrap.com/docs/5.0/content/typography/#headings) projects.

That aside, depending on how [we handle specificity of the Style System](https://github.com/ItsJonQ/g2/pull/284), `h1-h6` tagName selectors typically have some sort of global reset and possible top level styling. Some side-effects may result from that.

I think @diegohaz 's suggestion for using `role` is the best of both worlds:

```
<span role="heading" aria-level="1">
```